### DISCUSSION TITLE
Custom Transformation - AtlasMapUI

### DISCUSSION THREAD

**Discussion by sahilsethi12** [atlasmap/atlasmap#3481] [Q&A]
Hi ,

we are trying to import some custom jar into Atlas Map UI [2.2.3] , the mapping works fine in UI but gives a Null pointer exception while calling the atlasmap component  inside the code .
<img width="1792" alt="working_UI" src="https://user-images.githubusercontent.com/21212919/141790549-faf381db-64a9-4de6-8143-6b63aa01767d.png">

Steps to Reproduce the issue -->

1. Clone this Repository https://github.com/sahilsethi12/custom-java-trasnformer
2. Import existing adm file [ atlasmap-mapping-jsontoxml.adm , currently working to transform  json to xml without custom transformation ] .
3. Import the custom jar [ creating custom jar ----> Clone this repo https://github.com/sahilsethi12/custom-java-trasnformer and run mvn clean install and jar created inside directory upload the .jar in the UI ]
4. Once You Import the jar u can see the same java class inside the transformation drop down , map for any field and see the results by clicking  on the preview section and confirm the output [screenshot working_ui.png attached ]
5. Now click on export a new .adm file 
6. Now call this new custom .adm file inside the code and  check the error [already posted in read me file in the same repo]

**Comment by igarashitm** [comment]
Can you also upload the reproducer to run your mapping? through camel-atlasmap or invoking AtlasMap runtime API directly? What's the code exactly?

**Comment by sahilsethi12** [comment]
@igarashitm There's a camel route written :

I will update in Readme file for the same repository  , works ??

**Comment by sahilsethi12** [comment]
@igarashitm updated the code in  https://github.com/sahilsethi12/custom-java-trasnformer/blob/main/README.md

**Comment by igarashitm** [comment]
Nope not enough, please demonstrate how did you build your camel app and run it

**Comment by igarashitm** [comment]
One random guess, did you put custom transformation and service declaration files into the classpath of your camel app? it won't load from .adm file

**Comment by igarashitm** [comment]
Also this is fishy `Invalid serialized mapping definition content detected, discarding` - can you enable debug log for `io.atlasmap` and find more clue?

**Comment by sahilsethi12** [comment]
@igarashitm Please find my comments inline :

Nope not enough, please demonstrate how did you build your camel app and run it ---> Basically we are running camel app as a spring boot application , and running Application.java class 
You can find the complete code --> https://github.com/sahilsethi12/camel-atlasmap-poc

One random guess, did you put custom transformation and service declaration files into the classpath of your camel app? it won't load from .adm file --> yes custom transformation and service declaration files are not in classpath of camel application , What exactly  needs to add , can you please guide here ??

Also this is fishy Invalid serialized mapping definition content detected, discarding - can you enable debug log for io.atlasmap and find more clue? Sure will do that

**Reply by sahilsethi12** [reply]
By Adding Into camel app path we got the mapping , Thanks @igarashitm

**Reply by sahilsethi12** [reply]
@igarashitm One thing more , if i use AtlasMap version 2.3.2 , i don't see custom class in the drop down menu , any extra thing which needs to be added

**Comment by perfetimperfect24** [comment]
hi ,Does anyone has an update on this. I am unable to make custom transformation on atlasmap.

**Comment by gowthamnaraganti-cloud** [comment]
hi @sahilsethi12  do you have any update on custom transformation on atlasmap? unable to load in the function in transformations? dropdown
